### PR TITLE
gh-95601: restore support for awaitable objects that are not futures in `asyncio.wait`

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-95601

* Issue: https://github.com/python/cpython/issues/95920

### CODE REVIEW

File: Doc/library/asyncio-task.rst
Comment (graingert) [inline Doc/library/asyncio-task.rst]:
```suggestion
```

File: Lib/asyncio/tasks.py
Comment (kumaraditya303) [inline Lib/asyncio/tasks.py]: Use `warnings._deprecated`?

File: Lib/test/test_asyncio/test_tasks.py
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_tasks.py]: For new tests use `unittest.IsolatedAsyncTestCase`

File: Misc/NEWS.d/next/Library/2022-08-05-11-16-33.gh-issue-95601.hJvI9y.rst
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2022-08-05-11-16-33.gh-issue-95601.hJvI9y.rst]:
```suggestion
The explicit passing of awaitable objects that are not :class:`asyncio.Future` to :func:`asyncio.wait` is deprecated and will be removed in version 3.14.
```

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Lib/test/test_asyncio/test_tasks.py
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_tasks.py]: Also check the message with regex.

File: Lib/test/test_asyncio/test_tasks.py
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_tasks.py]:
```suggestion
                return asyncio.sleep(0).__await__()
```

File: Lib/test/test_asyncio/test_tasks.py
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_tasks.py]:
```suggestion
```

File: Lib/asyncio/tasks.py
Comment (graingert) [reply Lib/asyncio/tasks.py]: Yeah I usually would, but I kinda want to try backport the deprecation fix to 3.10

also the wording `warnings._deprecated` doesn't work too well for changes in behaviour and I didn't want to try and fix it and open a new can of worms there

File: Lib/test/test_asyncio/test_tasks.py
Comment (graingert) [reply Lib/test/test_asyncio/test_tasks.py]:
```suggestion
        with self.assertWarnsRegex(DeprecationWarning, "awaitable objects that are not futures"):
```

File: Lib/test/test_asyncio/test_tasks.py
Comment (graingert) [inline Lib/test/test_asyncio/test_tasks.py]:
```suggestion
        with (
            self.assertWarns(DeprecationWarning, "awaitable objects that are not futures"),
            self.assertRaises(TypeError),
        ):
```

File: Lib/test/test_asyncio/test_tasks.py
Comment (graingert) [inline Lib/test/test_asyncio/test_tasks.py]:
```suggestion
            self.assertWarnsRegex(
                DeprecationWarning, "awaitable objects that are not futures"
            ),
```

File: Lib/test/test_asyncio/test_tasks.py
Comment (kumaraditya303) [inline Lib/test/test_asyncio/test_tasks.py]:
```suggestion
```

**Review by kumaraditya303 — APPROVED** [review]
LGTM, I leave the decision of backporting to 3.10 to RM.

File: Lib/asyncio/tasks.py
Comment (graingert) [reply Lib/asyncio/tasks.py]: @hugovk are you already working on a plan to upbraid all the existing  `warnings.warn(..., DeprecationWarning)` to `warnings._deprecated` ?

there's a slight usability/wording concern when using it for changes in behavior instead of actual full removals
and there also seems to be a desire to use it for totally unscheduled removals: `remove=(4, )`

<hr>

in addition do you have a view on backporting deprecation, where it was clear from the code what the intention of the deprecation was, but the deprecation itself had a bug?

**Comment by graingert** [comment]
Looks like there's something worse going on here. It's possible to put a non-duck type future into wait and reach add_done_callback/remove_done_callback

File: Lib/asyncio/tasks.py
Comment (graingert) [inline Lib/asyncio/tasks.py:429]: this is getting added to the set with no checks whatsoever. It needs to check `_asyncio_future_blocking`

Comment (graingert) [reply Lib/asyncio/tasks.py:429]: fixed!

File: Lib/asyncio/tasks.py
Comment (hugovk) [reply Lib/asyncio/tasks.py]: > @hugovk are you already working on a plan to upbraid all the existing `warnings.warn(..., DeprecationWarning)` to `warnings._deprecated` ?

I'm not but that's a good idea. From a quick check, there's not too many, and most are set to be removed in 3.12 anyway.

> there's a slight usability/wording concern when using it for changes in behavior instead of actual full removals 

Can you alter the wording using the `message` parameter?

> and there also seems to be a desire to use it for totally unscheduled removals: `remove=(4, )`

Some removals scheduled for 4.0 were from before we knew if 3.10 or 4.0 would follow 3.9 (https://github.com/python/cpython/issues/80480#issuecomment-1093818166) and in at least one case is shorthand for "sometime after 2.7.final" (https://github.com/python/cpython/issues/80480#issuecomment-1093818167).

Anyway, we should review any remaining removals originally set for 4.0 (e.g. in new issues or relevant open issues) to decide if they can be removed in an upcoming 3.x.

---

> in addition do you have a view on backporting deprecation, where it was clear from the code what the intention of the deprecation was, but the deprecation itself had a bug?

Generally, I think it's a good idea to increase visibility and help users upgrade in time to reduce friction. It's good to ask release managers when in doubt.

Comment (graingert) [reply Lib/asyncio/tasks.py]: @hugovk do you think we have enough here to make a new issue out of this side discussion. Or maybe there's two to four ideas here:

* Do something about existing deprecation warnings that don't automatically get upgraded to errors
* Do something about the wording and format-string templating in warnings._deprecated to make it work for even the most subtle behaviour change
* Determine a backport strategy for when a deprecation warning was applied to the wrong thing
* Determine a strategy for permanent deprecations, where a feature is a bit broken or an alias of another thing, and people should really use the not broken way or source of the alias, but CPython is unwilling or undecided on actually removing it

Comment (hugovk) [reply Lib/asyncio/tasks.py]: Sure! Or at https://discuss.python.org/?

**Comment by kumaraditya303** [comment]
FWIW I don't think this should be deferred blocker. cc @pablogsal

File: Doc/library/asyncio-task.rst
Comment (pablogsal) [inline Doc/library/asyncio-task.rst]: We cannot deprecate behaviour after beta freeze. You need a steering council or release manager exception for this unfortunately

**Comment by pablogsal** [comment]
> FWIW I don't think this should be deferred blocker. cc @pablogsal 

If is adding a deprecation, is a blocker because we certainly won't be able to do it after the first release. It still needs an exemption though

File: Doc/library/asyncio-task.rst
Comment (graingert) [reply Doc/library/asyncio-task.rst]: This is restoring a behaviour that was supposed to be deprecated

Comment (pablogsal) [reply Doc/library/asyncio-task.rst]: Unfortunately, I don't follow. Could you give me some context on what happened?

**Comment by graingert** [comment]
> > FWIW I don't think this should be deferred blocker. cc @pablogsal
> 
> If is adding a deprecation, is a blocker because we certainly won't be able to do it after the first release. It still needs an exemption though

this is undoing a removal that didn't go via the deprecation policy, and then applying a deprecation for restored functionality

File: Doc/library/asyncio-task.rst
Comment (graingert) [reply Doc/library/asyncio-task.rst]: @pablogsal from the original issue report:

> this broke distributed and we missed the DeprecationWarning because ensure_future accepts a wider variety of objects than just futures tasks and coroutines
>
> https://github.com/dask/distributed/issues/6785
>
> the removal happened here https://github.com/python/cpython/commit/903f0a02c16240dc769a08c30e8d072a4fb09154

This PR restores support for Awaitable objects in `asyncio.wait` which was intended to be deprecated but never threw a DeprecationWarning because of an oversight in the implementation of the deprecation

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by gvanrossum** [comment]
Please don't add release-blocker labels without talking to Pablo first.

**Comment by graingert** [comment]
@tiran recommended spitting this PR into two, one that restores the functionality removed without deprecation and one that does the deprecation.

File: Misc/NEWS.d/next/Library/2022-08-05-11-16-33.gh-issue-95601.hJvI9y.rst
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-08-05-11-16-33.gh-issue-95601.hJvI9y.rst]:
```suggestion
Restore the support for the explicit passing of awaitable objects that are not :class:`asyncio.Future` to :func:`asyncio.wait`
```

**Comment by serhiy-storchaka** [comment]
It seems that the function was always supposed to only support futures and tasks. If you want to add support of arbitrary awaitables, it should be a new feature. And you need to update the documentation, the docstring and error messages to reflect this.

**Comment by graingert** [comment]
> It seems that the function was always supposed to only support futures and tasks. If you want to add support of arbitrary awaitables, it should be a new feature. And you need to update the documentation, the docstring and error messages to reflect this.

that's correct however support for Awaitables is currently available in 3.7, 3.8, 3.9, 3.10 and was removed as a consequence of an incorrect deprecation in 3.11

File: Lib/asyncio/tasks.py
Comment (serhiy-storchaka) [inline Lib/asyncio/tasks.py:420]: What will happened with implicitly created futures added to `new_fs`? Would it emit a warning?

Comment (graingert) [reply Lib/asyncio/tasks.py:420]: I removed the warning at @tiran 's request, with a plan to add it in a follow up PR.
I'm happy to put it back in this PR if you'd prefer

**Comment by graingert** [comment]
@serhiy-storchaka can you re-review this? I'll add the warning if you and @tiran agree it should be added in this PR.

**Review by gvanrossum — COMMENTED** [review]
I discussed this with @1st1 and a few others at the core dev sprint today, and we're not excited about this. There's definitely no support for trying to wedge this into the 3.11 release at this point.

IIRC @1st1 is planning to add his own response.

**Comment by kumaraditya303** [comment]
My previous reviews should be discarded as the code has changed since I reviewed. (I don't think I have the permission to dismiss stale reviews.). Since 3.11 is final now, it can't be backported either. I am hoping that @gvanrossum and @1st1 can handle this for 3.12.

**Comment by gvanrossum** [comment]
I think it’s better to start over with a clean slate.

**Pull request closed without merging**
